### PR TITLE
Fix: issue #32

### PR DESCRIPTION
**Pull request by CodingSinger**
1. modify registryProtocol as a singleton to fix graceful shutdown can't delete the provider url
2. fix notification confusion caused by multiple references sharing the same zk listener

### CODE REVIEW

File: registry/protocol/protocol.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ import (
 )
 
 var (
-	regProtocol *registryProtocol
+	regProtocol = newRegistryProtocol()
```
Comment (fangyincheng) [inline registry/protocol/protocol.go]: is `lazy loading` better for `newRegistryProtocol` ?

Comment (CodingSinger) [reply registry/protocol/protocol.go]: use the double-check to get or create the instance?

File: registry/zookeeper/listener.go [hunk 2]
```diff
@@ -37,18 +37,19 @@ import (
 
 // RegistryDataListener ...
 type RegistryDataListener struct {
-	interestedURL []*common.URL
-	listener      config_center.ConfigurationListener
+	subscribed map[*common.URL]config_center.ConfigurationListener
```
Comment (pantianying) [inline registry/zookeeper/listener.go]: is there Concurrency problem？

File: registry/protocol/protocol.go (hunk 1 above)
Comment (fangyincheng) [reply registry/protocol/protocol.go]: I think so , dubbo-check or lock or once?

File: registry/zookeeper/listener.go (hunk 2 above)
Comment (CodingSinger) [reply registry/zookeeper/listener.go]: i think it's unnecessary to consider Concurrency problem here.

File: registry/protocol/protocol.go (hunk 1 above)
Comment (CodingSinger) [reply registry/protocol/protocol.go]: done

**Comment by hxmhlt** [comment]
All registry's listener should be modified.

File: registry/zookeeper/listener.go (hunk 2 above)
Comment (AlexStocks) [reply registry/zookeeper/listener.go]: U should consider it.

File: registry/protocol/protocol.go [hunk 3]
```diff
@@ -348,10 +349,10 @@ func setProviderUrl(regURL *common.URL, providerURL *common.URL) {
 
 // GetProtocol ...
```
Comment (AlexStocks) [inline registry/protocol/protocol.go]: change the comment.

File: registry/protocol/protocol.go [hunk 4]
```diff
@@ -348,10 +349,10 @@ func setProviderUrl(regURL *common.URL, providerURL *common.URL) {
...
 func GetProtocol() protocol.Protocol {
```
Comment (AlexStocks) [inline registry/protocol/protocol.go]: Add comment.

File: registry/zookeeper/listener.go [hunk 5]
```diff
@@ -37,18 +37,26 @@ import (
 
 // RegistryDataListener ...
```
Comment (AlexStocks) [inline registry/zookeeper/listener.go]: change the comment, pls.

File: registry/zookeeper/listener.go [hunk 6]
```diff
@@ -37,18 +37,26 @@ import (
...
 type RegistryDataListener struct {
```
Comment (AlexStocks) [inline registry/zookeeper/listener.go]: Add comment .

File: registry/zookeeper/listener.go [hunk 7]
```diff
@@ -37,18 +37,26 @@ import (
...
-	interestedURL []*common.URL
-	listener      config_center.ConfigurationListener
+	subscribed map[*common.URL]config_center.ConfigurationListener
+	listener   config_center.ConfigurationListener
+	mutex      sync.Mutex
+	closed     bool
 }
 
 // NewRegistryDataListener ...
```
Comment (AlexStocks) [inline registry/zookeeper/listener.go]: change the comment.

File: registry/zookeeper/listener.go [hunk 8]
```diff
@@ -37,18 +37,26 @@ import (
...
+func (l *RegistryDataListener) SubscribeURL(url *common.URL, listener config_center.ConfigurationListener) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	if l.closed {
+		return
+	}
+	l.subscribed[url] = listener
 }
 
 // DataChange ...
```
Comment (AlexStocks) [inline registry/zookeeper/listener.go]: change the comment.

File: registry/zookeeper/listener.go [hunk 9]
```diff
@@ -81,6 +93,14 @@ func (l *RegistryDataListener) DataChange(eventType remoting.Event) bool {
 	return false
 }
 
+func (l *RegistryDataListener) Close() {
```
Comment (AlexStocks) [inline registry/zookeeper/listener.go]: add comment.

File: registry/zookeeper/registry.go [hunk 10]
```diff
@@ -155,8 +153,8 @@ func (r *zkRegistry) ZkClientLock() *sync.Mutex {
 }
 
 func (r *zkRegistry) CloseListener() {
```
Comment (AlexStocks) [inline registry/zookeeper/registry.go]: add comment.

File: registry/zookeeper/registry.go [hunk 11]
```diff
@@ -121,8 +120,7 @@ func newMockZkRegistry(url *common.URL, opts ...zookeeper.Option) (*zk.TestClust
 
 func (r *zkRegistry) InitListeners() {
 	r.listener = zookeeper.NewZkEventListener(r.client)
-	r.configListener = NewRegistryConfigurationListener(r.client, r)
-	r.dataListener = NewRegistryDataListener(r.configListener)
+	r.dataListener = NewRegistryDataListener()
 }
 
 func (r *zkRegistry) CreatePath(path string) error {
```
Comment (AlexStocks) [inline registry/zookeeper/registry.go]: add comment

File: registry/zookeeper/listener.go [hunk 12]
```diff
@@ -81,6 +93,14 @@ func (l *RegistryDataListener) DataChange(eventType remoting.Event) bool {
...
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	for _, listener := range l.subscribed {
+		listener.(*RegistryConfigurationListener).Close()
```
Comment (AlexStocks) [inline registry/zookeeper/listener.go]: pls close the listener outside of the lock scope.

File: registry/zookeeper/registry.go [hunk 13]
```diff
@@ -121,8 +120,7 @@ func newMockZkRegistry(url *common.URL, opts ...zookeeper.Option) (*zk.TestClust
 
 func (r *zkRegistry) InitListeners() {
```
Comment (AlexStocks) [inline registry/zookeeper/registry.go]: add comment for this func.

File: registry/zookeeper/listener.go [hunk 14]
```diff
@@ -37,18 +37,26 @@ import (
 
 // RegistryDataListener ...
 type RegistryDataListener struct {
-	interestedURL []*common.URL
-	listener      config_center.ConfigurationListener
+	subscribed map[*common.URL]config_center.ConfigurationListener
```
Comment (AlexStocks) [inline registry/zookeeper/listener.go]: @CodingSinger pls do not use pointer as map key.

**Comment by pantianying** [comment]
try zk reconnect case，there is a problem

File: registry/zookeeper/listener.go (hunk 14 above)
Comment (CodingSinger) [reply registry/zookeeper/listener.go]: why？ I see the golang spec includes a pointer-keyed map as an example.

File: registry/zookeeper/listener.go (hunk 5 above)
Comment (CodingSinger) [reply registry/zookeeper/listener.go]: ok

File: registry/zookeeper/listener.go (hunk 14 above)
Comment (AlexStocks) [reply registry/zookeeper/listener.go:40]: > why？ I see the golang spec includes a pointer-keyed map as an example.

show me the link.

Comment (AlexStocks) [reply registry/zookeeper/listener.go:40]: > why？ I see the golang spec includes a pointer-keyed map as an example.

ok, you win. by https://golang.org/ref/spec#Map_types.

Comment (CodingSinger) [reply registry/zookeeper/listener.go:40]: win-win

File: registry/base_registry.go [hunk 15]
```diff
@@ -178,7 +178,10 @@ func (r *BaseRegistry) RestartCallBack() bool {
 		}
 		logger.Infof("success to re-register service :%v", confIf.Key())
 	}
-	r.facadeBasedRegistry.InitListeners()
+
+	if flag {
+		r.facadeBasedRegistry.InitListeners()
+	}
```
Comment (Patrick0308) [inline registry/base_registry.go:184]: This flag will make the listener's pathMap of registry can not be cleaned every time.

Comment (AlexStocks) [reply registry/base_registry.go:184]: > This flag will make the listener's pathMap of registry can not be cleaned every time.

good job.

**Comment by pantianying** [comment]
I tested your code, but I still couldn't find the provider

File: registry/base_registry.go (hunk 15 above)
Comment (CodingSinger) [reply registry/base_registry.go:184]: > This flag will make the listener's pathMap of registry can not be cleaned every time.

this is not  the main problem

**Pull request merged**